### ISSUE TITLE
Window size/position different on Win10, Linux/X11, and Linux/Wayland

### ISSUE THREAD

**Issue by PaulSquires** [cycfi/elements#422]
I used the code from the text_edit example only modifying the window creation call. (I used the Master branch).

```
int main(int argc, char* argv[])
{
   app _app("TextEdit");
   window _win(_app.name(), window::standard, rect(300,300,1000,500));
   _win.on_close = [&_app]() { _app.stop(); };

   view view_(_win);

   view_.content(
      make_edit_box(),
      make_bkd()
   );

   _app.run();
   return 0;

}
```

As you can see, the only modification is:
```
window _win(_app.name(), window::standard, rect(300,300,1000,500));
```
I compiled and and ran the code on 3 different systems:
(1) Windows 10 (64-bit) using Visual Studio 2022.
(2) Fedora Workstation 40 (Gnome + X11) using CMake.
(3) Fedora Workstation 40 (KDE Plasma 6.1.4 + Wayland) using CMake.

All 3 seem to produce different x/y positioning and xx/yy sizing.

**(1) Windows 10 (64-bit) using Visual Studio 2022.**
![windows_10 (1)](https://github.com/user-attachments/assets/d6467dc7-d5c0-4047-8e75-899cc7666c95)

**(2) Fedora Workstation 40 (Gnome + X11).**
![linux_x11 (1)](https://github.com/user-attachments/assets/728aab0b-c408-4c1b-84f9-15a9ab2e178f)

**(3) Fedora Workstation 40 (KDE Plasma 6.1.4 + Wayland).**
![linux_wayland (1)](https://github.com/user-attachments/assets/b9e6b687-32af-4700-a0ef-d25328c57a26)

Must be something with how the underlying host windowing frameworks handle the creation of the windows by applying their own opinionated settings and disregarding the values used by Elements (?)

I am **very** new to trying the Elements framework so this discussion may have come up in the past.

I am willing to test whatever suggestions you provide.

**Comment by djowel** [comment]
Hmmm... Looks like I need to do some tweaking on these platforms to see what's going on. It's most likely a problem with Element's windows handling on Linux, I'm not sure. I typically just use the defaults, without explicit sizing. And most of the time, I use elements for making plugins where the host is the one setting up the window for you.

You are welcome to do some sleuthing too, if  you will :-)

**Comment by PaulSquires** [comment]
Lol, I am sleuthing and I will report back here once I find (hopefully) a solution. I am not a Gtk+ expert by any means but so far the documentation and concepts have been pretty easy to follow.

**Comment by PaulSquires** [comment]
It does seem like the Wayland compositor is the problem.

Here is a pretty succinct answer from a trusty A.I. response that summarizes everything I've experienced.

> gtk_window_move wayland
>
> Based on the search results, it appears that moving a window using GTK3 on Wayland is not directly possible due to limitations in the Wayland protocol. The gtk_window_move function, which works on Xorg, does not work on Wayland.
> 
> The issue is not specific to GTK3, but rather a fundamental limitation of Wayland. According to the Wayland documentation, the wl_surface interface does not provide a way to move a surface (window) to a different position. Instead, the compositor (window manager) is responsible for managing window placement and movement.
> 
> In the provided Stack Overflow question, the user tried various approaches, including setting type hints and transient parents, but none of these worked on Wayland. Another user suggested trying a different window manager, implying that the issue might be specific to the current compositor.
> 
> If you still want to achieve window movement on Wayland, you could consider using a different toolkit or framework that provides this functionality, such as Qt5 with the -platform wayland option (as mentioned in the Stack Overflow question). However, this would require rewriting your application to use a different toolkit.
> 
> In summary, due to the limitations of the Wayland protocol, it is not currently possible to move a window using GTK3 on Wayland. You may need to consider alternative approaches or toolkits if window movement is a requirement for your application.
> 
> Here is an example of the relevant GTK3 code snippet, highlighting the issue:
> 
> ```
> // Assuming gtk_window is a valid GTK3 window object
> gtk_window_move(my_window, 100, 200);
> ```
> This code will not work on Wayland, as the gtk_window_move function relies on Xorg-specific functionality.
>  

Other than switching the backend host to something other than Gtk, I doubt that this can be fixed.

**Comment by PaulSquires** [comment]
I have not investigated any potential issues related to the initial **sizing** of a window. Hopefully there is nothing wrong with that but I'll investigate it for the sake of completeness.

**Comment by PaulSquires** [comment]
Sizing appears to be fine. 

The inability to position the window's startup position under Wayland is not a deal breaker for me especially now that I know the expected behaviour under Linux/X11, Linux/Wayland, and Windows. I don't own or use Apple so I can not comment on any behaviour on that system.

Should be okay to close this issue now.

**Comment by djowel** [comment]
> Sizing appears to be fine.
> 
> The inability to position the window's startup position under Wayland is not a deal breaker for me especially now that I know the expected behaviour under Linux/X11, Linux/Wayland, and Windows. I don't own or use Apple so I can not comment on any behaviour on that system.
> 
> Should be okay to close this issue now.

Interesting. It's odd why wayland does not support such a thing. 

My main work computer is the Mac, so no worries there. I'll close this now. Thanks for investigating on this! Much appreciated!

**Comment by djowel** [comment]
> Other than switching the backend host to something other than Gtk, I doubt that this can be fixed.

This I do not understand. If Wayland is the problem, then how can other backends work around this, and not Gtk? Gtk is just a layer on top of Wayland, like others.

**Comment by PaulSquires** [comment]
> Other than switching the backend host to something other than Gtk, I doubt that this can be fixed.

Yes, I should not have said that because it certainly makes no sense at all.  Lol. 
Wayland is the problem, not the chosen backend, so until Wayland changes the problem will persist.

**Comment by djowel** [comment]
Feel free to close this, if you are satisfied with the conclusion. OTOH, we might also want to investigate further. I'm naturally skeptical of A.I. responses, I'm sure you know what I mean :-)

**Comment by PaulSquires** [comment]
Understood. I did a lot of testing by modifying the existing the Gtk3 hosts directory files to include things like various Gtk hints. I even thought that maybe the timing of when the deferred on_activate vector was processed may have interfered with the positioning. It didn't (of course), and it is cruel that Gtk made you have to resort to such gymnastics. I'll close the issue for now but should any additional info come to light then I'll ask to reopen, etc.

**Comment by djowel** [comment]
> Understood. I did a lot of testing by modifying the existing the Gtk3 hosts directory files to include things like various Gtk hints. I even thought that maybe the timing of when the deferred on_activate vector was processed may have interfered with the positioning. It didn't (of course), and it is cruel that Gtk made you have to resort to such gymnastics. I'll close the issue for now but should any additional info come to light then I'll ask to reopen, etc.

I applaud your efforts! It's not common that people go to that extent, actually, beyond the superficial inquiries. SO, much appreciated!

I am also considering other "hosts" in Linux, even perhaps piggyback on QT, direct to X11, etc. Might you be interested in that angle, do tell.

**Comment by PaulSquires** [comment]
Oh yes, I am interested for sure. I am enjoying my journey in finding a robust cross platform application solution that isn't mega heavy and bloated. Don't bother with X11 though because it's going the way of the Doodoo bird sooner rather than later, so any effort into support X11 is probably not fruitful. 

I would even be up for creating more user controls like a calendar, date picker, etc. I know that I have the skills for that kind of stuff because I wrote them from scratch using Windows API... I just need to get better acquainted with the Elements way of constructing things.

I will end on this note that I forgot to mention earlier and which probably doesn't deserve a pull request (which I'd have to learn how to do as I host open source projects but have never had to create a pull request for someone else's project).  :-)

When compiling and testing this Wayland issue, GCC gave a compiler warning emanating from the Gtk library that a constant was deprecated. 

Basically change **G_APPLICATION_FLAGS_NONE** to **G_APPLICATION_DEFAULT_FLAGS**

```
   struct init_app
   {
      init_app()
      {
//         the_app = gtk_application_new(nullptr, G_APPLICATION_FLAGS_NONE);
         
         //NOTE: G_APPLICATION_FLAGS_NONE is deprecated use G_APPLICATION_DEFAULT_FLAGS
         
         the_app = gtk_application_new(nullptr, G_APPLICATION_DEFAULT_FLAGS);
         g_signal_connect(the_app, "activate", G_CALLBACK(activate), nullptr);
      }
   };
```

Seems like a simple fix.  :-)

**Comment by djowel** [comment]
> Basically change **G_APPLICATION_FLAGS_NONE** to **G_APPLICATION_DEFAULT_FLAGS**

Thanks! Will do.
